### ISSUE TITLE
Data validity for audience tiles when connected to Analytics

### ISSUE THREAD

**Issue by kelvinballoo** [google/site-kit-wp#8763]
## Bug Description

I don't think there is a task but we should absolutely look into data validity for those data that we surface on audience tiles for audience segmentation. 

If you look at the screenshots below, you will notice that for cities, we have percentages like 119400% which is very strange.
Moreover, we'd have to list down the logic from where or how we are computing certain data from analytics.
E.g.  how are we getting the visits per visitor? On analytics, I don't see that number and chances are we are computing it?

Note: This ticket stemmed from the comment here: https://github.com/google/site-kit-wp/issues/8138#issuecomment-2137201828 

## Steps to reproduce
1. Set up Audience tiles
2. Connect oi.ie in analytics
3. Look at the tiles data

## Screenshots

<img width="1115" alt="Screenshot 2024-05-29 at 14 33 40" src="https://github.com/google/site-kit-wp/assets/125576926/d1bde76e-a5d0-4b4a-b2d9-ebb3e8cbf6b2">

<img width="964" alt="Screenshot 2024-05-29 at 14 39 08" src="https://github.com/google/site-kit-wp/assets/125576926/bb91ae59-1c49-463d-9296-e31414d0e8c3">

## Additional Context

 - PHP Version: 8.0
 - Browser: Chrome
 - Plugin Version: 1.127.0
 - Device: Macbook Pro 16inch MacOS Sonoma

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* In order to provide a consistent view of the following metrics:
  * The "pages per visit" metric in an Audience Tile should be formatted to two decimal places, to match the format of the metric in the corresponding Key Metric tile.
  * The "cities with the most visitors" metric values in an Audience Tile should be calculated by dividing the visitor count for the city by the total visitor count (both values being per audience) in order to provide the correct percentage value.
* When viewing the "all visitors" Audience Tile, the above metrics should match the "pages per visit" and "top cities driving traffic" Key Metrics tiles.

## Implementation Brief

- [x] In `assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTile/AudienceTileMetric.js`
  - [x] Accept a new prop `metricValueFormat`, this would be the object which should be passed to `numFmt` function as `options`.
- [x] In `assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTile/index.js`
  - [x] For "Pages per visit" `AudienceTileMetric` component, pass `metricValueFormat` prop as:

  ```js
    {
        style: 'decimal',
        maximumFractionDigits: 2,
    }
  ```

- [x] In `assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTile/AudienceTileCitiesMetric.js`
  - [x] Divide `topCities?.metricValues[ index ]?.value` by `topCities.total` to get the correct percent value.

### Test coverage

* Update stories for `AudienceTile` to display correct metric values.

## QA Brief

* Confirm the metrics in the bug description are now shown and calculated correctly.

## Changelog entry

* N/A

**Comment by techanvil** [comment]
Hi @kelvinballoo, thanks for raising this issue. The "cities" metric is clearly incorrect and it's entirely reasonable to question the validity of the metrics.

To your question about calculating metrics, the only metric we're actually calculating (as opposed to simply rounding) is in fact the "cities" one - and in fact the problem observed is because we're _not_ performing the expected calculation (we're currently not dividing the values for the "cities" metric by the total number of visitors).

The metrics that are used in the reports are defined in the [design doc](https://docs.google.com/document/d/1MGD5Djy6AeeZC4zBtHqS-lQEWD9jw0kf-IIWw-jLCFU/edit#heading=h.lsud3imhtyov). As you'll see, "visits per visitor" is directly retrieved via the `sessionsPerUser` metric, and the other metrics barring "cities" have similiarly direct mappings.

You may notice some discrepancies between data in Site Kit and reports/explorations in Analytics. The difference in the headline "visitors" figure ("users" in Analytics) is due to Site Kit using the `totalUsers` metric whereas the Analytics report uses `activeUsers`. We have a couple of open issues for this - #7214 and #7263.

If you notice other differences it's quite possible they can be explained by the factors discussed in this comment: https://github.com/google/site-kit-wp/issues/7214#issuecomment-1674687228.

An additional aspect to consider on the topic of data consistency is how we present the same metrics in multiple places within Site Kit itself. It's worth noting that all of the metrics for an audience are viewable elsewhere in Site Kit, so when viewing the "all visitors" audience, we can cross reference other widgets as a sanity check for the figures.

To illustrate the point, here's a map of the metrics on the Audience Tile to the same metrics on Key Metric tiles and the All Traffic widget:

![image](https://github.com/google/site-kit-wp/assets/18395600/2e7b7d41-1ed0-454c-86c6-2c12d863f0ef)

As you can see most of them match up. The "cities" values are totally off, and there are a couple of additional differences which are due to rounding.

The net of this is that we need to fix the "cities" values, and we should provide consistent rounding where it makes sense to - specifically, I think the "pages per visit" metric should be rounded to two decimal places on the Audience Tile to match the KM tile. The "total pageviews" figure can remain rounded as it is, in order to match the formatting of the "visitors" figure, it would look a bit odd otherwise. I have therefore added AC to this issue to address these specific points.

Hopefully that clarifies things and provides some useful background / context. It's undoubtedly a complex topic. Please let me know if there's anything that could use further clarification or discussion! :)

**Comment by techanvil** [comment]
Hey @ankitrox, thanks for drafting this IB. A couple of points:

Considering that `numFmt()` is already called on `metricValue` in `AudienceTileMetric`, the suggested approach would result in `numFmt()` being called twice on the "pages per visit" value.
https://github.com/google/site-kit-wp/blob/721401ea2fac18fad433bf2e066aaa486aa83e79/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTile/AudienceTileMetric.js#L42

I'd suggest it would be slightly cleaner to take a similar approach to `MetricTileNumeric`, passing a `metricValueFormat` prop into `AudienceTileMetric`.

https://github.com/google/site-kit-wp/blob/721401ea2fac18fad433bf2e066aaa486aa83e79/assets/js/components/KeyMetrics/MetricTileNumeric.js#L31-L48

Also, `pagesPerVisit.metricValueFormat` is not defined. We should be copying the format from `PagesPerVisitWidget`, and if taking the above approach we can pass it directly to `AudienceTileMetric` via the `metricValueFormat` prop.

https://github.com/google/site-kit-wp/blob/721401ea2fac18fad433bf2e066aaa486aa83e79/assets/js/modules/analytics-4/components/widgets/PagesPerVisitWidget.js#L103-L106

**Comment by techanvil** [comment]
Thanks @ankitrox! The IB LGTM. :white_check_mark:

**Comment by benbowler** [comment]
FYI @techanvil, @mohitwp above found a bug where data was inconsistent on the "All visitors" tile. I found a clear way to repeat this:

1. Set just the "All visitors" tile in the selection panel, you will see the correct stats:
<img width="985" alt="Screenshot 2024-07-11 at 12 51 14" src="https://github.com/google/site-kit-wp/assets/119352/3c1bf9d5-a7fb-4b18-af86-0c11d1f64d73">

2. Then add back in the "Returning visitors" tile, you will see incorrect/mixed stats:
<img width="975" alt="Screenshot 2024-07-11 at 13 08 20" src="https://github.com/google/site-kit-wp/assets/119352/17ee4ed1-ddfd-4d86-b348-e28e4e590ce9">

The reason this caused mixed metrics was because these steps would show the tiles in a different order to their default order and the order in which the audiences are queries and returned in the GA reports. I fixed this in [this PR](https://github.com/google/site-kit-wp/pull/9003) by correctly filtering rows rather than relying on their order.

As this is new work I will put it into CR.

**Comment by techanvil** [comment]
Hi @mohitwp and @benbowler. Thanks for both spotting this and finding a fix; however - it's actually a known issue, https://github.com/google/site-kit-wp/issues/8921, and that issue has an IB ready to review and a draft PR.

Actually, I do prefer the PR for that issue - it's very similar but it also looks up the rows by the `dateRange` dimension which adds an extra element of safety.

As such, and with a view to ensuring this issue doesn't creep over the estimate, I'd say let's keep this issue to what's defined and address the above defect via #8921.

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on main environment.
- Verified the "pages per visit" metric in an Audience Tile  formatted to two decimal places, to match the format of the metric in the corresponding Key Metric tile.
- Verified when viewing the "all visitors" Audience Tile, the metrics  match the "pages per visit" and "top cities driving traffic" Key Metrics tiles.
- Verified audience tiles and key metrics tiles data are matching for all dates.

**Note : Data is still not showing correct in Audience tiles as mentioned above. As per @techanvil this issue will be fix by #8921.**
Also, I've notice analytics shows different data city wise. I will open a different ticket for discussion because as per AC Audience tile matches the key metrics data. 



**Last 28 days**

<details>

![image](https://github.com/user-attachments/assets/6fad638f-2874-4a5a-b0a9-556f20ff7709)

![image](https://github.com/user-attachments/assets/412655ee-e2a8-408a-8f26-c51724cfc000)

</details>


**Last 7 Days** 

<details>

![image](https://github.com/user-attachments/assets/03a9fa50-ff5d-49fb-8f07-96552d19cfc9)

![image](https://github.com/user-attachments/assets/214ba4a0-22f9-4e39-85e4-80c6e60919ae)

</details>

**Last 14 days**

<details>

![image](https://github.com/user-attachments/assets/1c2bf4f1-cbe8-466e-acaf-a2dc25a8f363)

![image](https://github.com/user-attachments/assets/058c1f51-dd2e-4545-8270-fe66aef0f93d)

</details>

**Last 90 Days**

<details>

![image](https://github.com/user-attachments/assets/8802a823-254d-425c-b92e-910d93f144fa)

![image](https://github.com/user-attachments/assets/84d375a8-3957-49df-8574-ce3396fdcee4)


</details>
